src/main/java/javaslang/collection/HashSet.java

```diff
@@ -481,11 +481,12 @@ public int hashCode() {
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public boolean equals(Object o) {
         if (o == this) {
             return true;
         } else if (o instanceof HashSet) {
-            final HashSet<?> that = (HashSet<?>) o;
+            final HashSet<T> that = (HashSet<T>) o;
```

Comment:
I tested a little bit and recognized that we need another equals implementation.

The underlying HashArrayMappedTrie may have stored the elements in an arbitrary order, so that the iterators do not return the elements in a specific order.

This lead to this implementation:

```java
@SuppressWarnings("unchecked")
@Override
public boolean equals(Object o) {
    if (o == this) {
        return true;
    } else if (o instanceof HashSet) {
        final HashSet<?> that = (HashSet<?>) o;
        return this.length() == that.length() && ((HashSet<Object>) this).containsAll(that);
    } else {
        return false;
    }
}
```

We do not cast to `<T>` in order to prevent ClassCastExceptions.

Thanks for for your finding!!